core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -0,0 +1,259 @@
+                }
+                switch (state) {
+                    case REQUIRE_HEAD: {
+                        sendHead();
+                        continue;
+                    }
+                    case REQUIRE_BODY: {
+                        if (!subscribed) {
+                            subscribed = true;
+                            publisher.subscribe(this);
+                            return;
+                        }
+                        if (upstream != null) {
+                           requestUpstream(upstream);
+                        }
+                        return;
+                    }
+                    case REQUIRE_TAIL: {
+                        sendTail();
+                        continue;
+                    }
+                    case REQUIRE_COMPLETE: {
+                        sendComplete();
+                        return;
+                    }
+                    case DONE: {
+                        upstream.cancel();
+                        return;
+                    }
+                }
+            }
+        }
+
+        private void sendHead() {
+            setState(State.REQUIRE_HEAD, State.REQUIRE_BODY);
+            requestedUpdater.decrementAndGet(this);
+            downstream.onNext(head);
+        }
+
+        private void sendTail() {
+            setState(State.REQUIRE_TAIL, State.REQUIRE_COMPLETE);
+            requestedUpdater.decrementAndGet(this);
+            if (tail != null) {
+                downstream.onNext(tail);
+            } else {
+                sendComplete();
+            }
+        }
+
+        private void sendComplete() {
+            setState(State.REQUIRE_COMPLETE, State.DONE);
+            requestedUpdater.decrementAndGet(this);
+            downstream.onComplete();
+        }
+
+        private void requestUpstream(Subscription subscription) {
+            for (;;) {
+                final long requested = this.requested;
+                if (requested == 0) {
+                    return;
+                }
+                if (requestedUpdater.compareAndSet(this, requested, 0)) {
+                    for (;;) {
+                        final long oldNeedToPublish = needToPublish;
+                        final long newNeedToPublish = LongMath.saturatedAdd(oldNeedToPublish, requested);
+                        if (needToPublishUpdater.compareAndSet(this, oldNeedToPublish, newNeedToPublish)) {
+                            break;
+                        }
+                    }
+                    subscription.request(requested);
+                    return;
+                }
+            }
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            if (cancelled) {
+                subscription.cancel();
+                return;
+            }
+            upstream = subscription;
+            requestUpstream(subscription);
```

Comment:
This could break the internal state, for example:
- Thread A requests 1 element to `Subscription`.
- `requested` is updated to 1 by Thread A.
- Thread A stays in L132 to deliver `head`.
- Thread B invokes `onSubscribe()` `requestUpstream(subscription)`
- Thread B updates `requested` to 0 in L190.
- Thread A decrease `requested` by 1 in L164.
- `requested` becomes -1.

Reply:
✅ fixed by extends `StreamMessage` and change `compareAndSet(requested, 0)` to `requested--`